docker/bookkeeper/entrypoint.sh

```diff
@@ -48,31 +56,151 @@ export BK_tlsTrustStoreType=JKS
-else
-  # The container does not contain any BK data, it is probably a new
-  # bookie. We will format any pre-existent data and metadata before starting
-  # the bookie to avoid potential conflicts.
-  echo "format bookie data and metadata"
-  /opt/bookkeeper/bin/bookkeeper shell bookieformat -nonInteractive -force -deleteCookie
-fi
-
-echo "start bookie"
-/opt/bookkeeper/scripts/entrypoint.sh bookie
+# Create directories for multiple ledgers and journals if specified.
+create_bookie_dirs() {
+  IFS=',' read -ra directories <<< $1
+  for i in "${directories[@]}"
+  do
+      mkdir -p $i
+      if [ "$(id -u)" = '0' ]; then
+          chown -R "${BK_USER}:${BK_USER}" $i
+      fi
+  done
+}
+
+wait_for_zookeeper() {
+    echo "Waiting for zookeeper"
+    until zk-shell --run-once "ls /" ${BK_zkServers}; do sleep 5; done
+}
+
+
+create_zk_root() {
+    if [ "x${BK_CLUSTER_ROOT_PATH}" != "x" ]; then
+        echo "Creating the zk root dir '${BK_CLUSTER_ROOT_PATH}' at '${BK_zkServers}'"
+        zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH} '' false false true" ${BK_zkServers}
+    fi
+}
+
+configure_bk() {
+    # We need to update the metadata endpoint and Bookie ID before attempting to delete the cookie
+    sed -i "s|.*metadataServiceUri=.*\$|metadataServiceUri=${BK_metadataServiceUri}|" /opt/bookkeeper/conf/bk_server.conf
+    if [ ! -z "$BK_useHostNameAsBookieID" ]; then
+      sed -i "s|.*useHostNameAsBookieID=.*\$|useHostNameAsBookieID=${BK_useHostNameAsBookieID}|" ${BK_HOME}/conf/bk_server.conf
+    fi
+}
+
+fix_bk_ipv6_check() {
+
+   # `${SCRIPTS_DIR}/common.sh` - a shell script in Bookeeper, runs this command: `/sbin/sysctl -n net.ipv6.bindv6only`.
+   # This command can return an error exit code `255` and error message:
+   #    `sysctl: cannot stat /proc/sys/net/ipv6/bindv6only: No such file or directory`
+   #
+   # If the `common.sh` file is sourced in a script that sets "exit on error" as true (set -e), the script shall return
+   # an error code `255`. Since, the `bookkeeper/bin/bookeeper` shell script sources the `common.sh` file, and it sets
+   # "exit on error" as true, we encounter this error when it is invoked. To avoid that error, here we modify the
+   # `common.sh` file to execute `/sbin/sysctl -n net.ipv6.bindv6only` only if file `/proc/sys/net/ipv6/bindv6only`
+   # is available.
+
+   # Replace the 22nd line with "if [ -f /sbin/sysctl ] && [ -f /proc/sys/net/ipv6/bindv6only ]; then".
+   sed -i "22s|.*|if [ -f /sbin/sysctl ] \&\& [ -f /proc/sys/net/ipv6/bindv6only ]; then|" ${BK_HOME}/bin/common.sh
+}
+
+initialize_cluster() {
+    set +e
+
+    zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+    if [ $? -eq 0 ]; then
+        echo "Cluster metadata already exists"
+    else
+        # Create an ephemeral zk node `bkInitLock` for use as a lock.
+        lock=`zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH}/bkInitLock '' true false false" ${BK_zkServers}`
```

Comment:
I don't understand this. If the script is set to run once, then I interpret it as it is not going to hold a zk session. If it is not holding a bk session, then all ephemerals will be deleted upon the script completing its execution and there will be no lock whatsoever. This smells like a bug.

Also, where is unlock being implemented? For an ephemeral lock znode, either it is deleted explicitly or it is gone when the session closes. Which one is happening here?

Reply:
The idea here is that if there are multiple containers coming up simultaneously, one of them reaches here, holds the lock and performs bookkeeper init cluster, while the rest of the containers wait for the initialization to get completed. 

To answer your question: the latter is happening. 

To provide a bit more background: The same mechanism is present in `apache/bookkeeper:4.7.3`: https://github.com/apache/bookkeeper/blob/f307a36acf9d2acf5c3ff315e1b3bfd7873e3cce/docker/scripts/init_bookie.sh#L43. 

In `apache/bookkeeper: 4.8.2`, the code uses `bookkeeper org.apache.zookeeper.ZooKeeperMain` command, but that command doesn't work, so we had to bring in `zk-shell` back in our implementation. 
https://github.com/apache/bookkeeper/blob/4fa14617a855df236a902e76c8c68a21952bb48a/docker/scripts/init_bookie.sh#L50

Reply:
> The idea here is that if there are multiple containers coming up simultaneously, one of them reaches here, holds the lock and performs bookkeeper init cluster, while the rest of the containers wait for the initialization to get completed.

That's not what I'm asking, please read my comment again.

> To answer your question: the latter is happening.

The latter what? If the shell is running in non-interactive mode (`run-once`), then it is not holding a session, which means that ephemerals are gone and the lock is not locking anything.

It seems broken to me.

Reply:
I ran a quick experiment to test whether the logic works or not. The minor changes I did for debugging were:
* Added "set -x" to the script so that all commands are printed in the logs
* Added an additional `echo` statement to print the lock (as shown below)
* Added an additional `zk-shell ls` command (as shown below) to check if the `bkInitLock` is seen by nodes which don't create it.

```
...
 # Create an ephemeral zk node `bkInitLock` for use as a lock.
 lock=`zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH}/bkInitLock '' true false false" ${BK_zkServers}`
 echo "Lock: $lock" <--- Added this line 
 zk-shell --run-once "ls ${BK_CLUSTER_ROOT_PATH}/" ${BK_zkServers} <--- Added this line
...
```
The behavior I found upon looking at the logs seems correct: one node  (BK 2) creates the `/pravega/pravega-cluster/bookkeeper/bkInitLock` path and initializes the cluster, and the others (BK 1 and BK 3) find that it is already present and don't initialize the cluster. 

```
root@pvgcompose:~/bk_upgrade_pr# grep "Lock:" *.log
bk1_23-13-15--2020-01-21.log:Lock: Path /pravega/pravega-cluster/bookkeeper/bkInitLock already exists
bk2_23-13-15--2020-01-21.log:Lock:
bk3_23-13-15--2020-01-21.log:Lock: Path /pravega/pravega-cluster/bookkeeper/bkInitLock already exists
```

The node that creates the lock has the following in the logs: 
```
Lock:
bkInitLock
Bookkeeper znodes do not exist in Zookeeper. Initializing a new Bookeekeper cluster.
initnewcluster operation succeeded
```
The nodes that find the lock is already established have this in the logs: 
```
Lock: Path /pravega/pravega-cluster/bookkeeper/bkInitLock already exists
bkInitLock
Another instance might be initializing the cluster at the same time.
```

So, the logic seems to be working fine here, from what I can see. This piece of code is same as what you can see in `apache/bookkeeper` code here: 
https://github.com/apache/bookkeeper/blob/2fb7f3a80561eb70dc063c89763d41519fb45f1e/docker/scripts/init_bookie.sh#L43

Reply:
Even though this is used in the BK code, I'm not convinced this is entirely correct. One additional point to support what I'm saying is that neither the `bkInitLock` znode is deleted nor a session is closed explicitly.

To recap the semantics of ephemerals, a znode exists for as long as its session is considered alive. It is possible that the session within which the znode was created was not closed and the ephemeral is not deleted while it is alive, but it will eventually timeout, and if it times out before we complete the operation, then the lock will fail. If for some reason we have to acquire the lock again within a short period, then the creation will fail incorrectly.

We need to be able to acquire and release explicitly, and we need to make sure that we hold the session for the ephemeral while the lock is acquired. I'm not up-to-speed with the syntax and semantics of `zk-shell`, perhaps we can have a look to see how to implement this correctly?

Reply:
@fpj would it be OK to explore what you have suggested as part of a separate issue, especially given the fact that this code has been working for us via the `apache/bookkeeper` `4.7.3` image code previously and now in this `pravega/bookkeeper` image as seen via manual tests and automated system tests? I haven't been able to find good documentation on `zk-shell` and believe it'll require some experimentation to understand the behavior of its commands precisely. What do you say?

Reply:
I don't have a definitive answer right now and am digging further. I've found that the `bkInitLock` is indeed an ephemeral znode that goes away shortly after the node is created. So, there is certainly a potential for a race condition causing issues here. Copying the result of the experiment here (for reference): 

```
# Command format: create <path> <value> [ephemeral] [sequence] [recursive] [async]
# All [..] are false by default
$ zk-shell --run-once 'create /test1 '\'''\'' false false false' zookeeper:2181
$ zk-shell --run-once 'ls /' zookeeper:2181
test1
$ zk-shell --run-once 'ls /test1' zookeeper:2181
[empty]
$ zk-shell --run-once 'create /test1/bkInitLock '\'''\'' true false false' zookeeper:2181
$ zk-shell --run-once 'ls /test1/' zookeeper:2181
bkInitLock
$ zk-shell --run-once 'ls /test1/' zookeeper:2181
bkInitLock
$ zk-shell --run-once 'ls /test1/' zookeeper:2181
bkInitLock
$ zk-shell --run-once 'ls /test1/' zookeeper:2181
bkInitLock
$ zk-shell --run-once 'ls /test1/' zookeeper:2181
[empty]
```

Reply:
Confirming that system tests have passed with these changes. The run was: `test-pravega-continuous/2589/` and it comprised of auth-enabled Pravega cluster.

Reply:
This conversation continued here: https://github.com/pravega/pravega/pull/4420#discussion_r370048900.